File: Java/src/stocks/Transaction.java
package stocks;

import java.awt.Color;
import java.awt.Label;

import exceptions.CouldntFindInfoAboutStock;

public class Transaction {
	
	private static int ID;
	private int id;
	private Stock stock;
	private int quantity;
	private double price;
	
	public Transaction(Stock stock, int quantity) throws CouldntFindInfoAboutStock {
		this.stock = stock;
		this.quantity = quantity;
		this.price = getCurrentPrice();
		this.id = ++ID;
	}

	public Transaction(Stock stock, int quantity, int id, double price) {
		this.id = id;
		this.stock = stock;
		this.quantity = quantity;
		this.price = price;
		ID = id;
	}

	public double getCurrentPrice() throws CouldntFindInfoAboutStock {
		String timestamp = "1616072670";
		Crawler crawler = new Crawler();
		crawler.makeJSON(stock.getName(), timestamp, timestamp);
		JSONParser jsp = new JSONParser("files/" + stock.getName() + ".json");
		return jsp.getCurrentPrice();
	}
	
	double getRelativeDiff(double x, double y) {
		if (x > y) return (x / y - 1) * 100;
		else return (y / x - 1) * 100;
	}
	
	public Label toLabel() {
		try {
			Label label = new Label();
			double currentPrice = getCurrentPrice();
			if (price - currentPrice >= 0) label.setForeground(Color.RED);
			else label.setForeground(Color.decode("#36b622"));
			label.setText(String.format("%-3d %-8s %-8d   %-13.2f   %-13.2f   %-17.2f   %-17.2f", id, stock.getName(), quantity, price, currentPrice, 
					Math.abs(currentPrice - price), getRelativeDiff(currentPrice, price)));
			return label;
		} catch (CouldntFindInfoAboutStock e) { return null; } 
	}

	public int getId() {
		return id;
	}

	public int getQuantity() {
		return quantity;
	}

	public double getPrice() {
		return price;
	}
	
	public Stock getStock() {
		return stock;
	}

	public static int getID() {
		return ID;
	}

	public static void setID(int iD) {
		ID = iD;
	}

	public void setQuantity(int quantity) {
		this.quantity = quantity;
	}
}

File: C++/transaction.cpp
#include "transaction.h"

int Transaction:: ID = 1;

double Transaction::getRelativeDiff(double x, double y) const {
	if (x > y) return (x / y - 1) * 100;
	else return (y / x - 1) * 100;
}

void Transaction::copy(const Transaction & t) {
	this->id = t.id;
	this->stock = new Stock(*(t.stock));
	this->quantity = t.quantity;
	this->price = t.price;
}

void Transaction::move(Transaction& t) {
	this->id = t.id;
	this->stock = t.stock;
	t.stock = nullptr;
	this->quantity = t.quantity;
	this->price = t.price;
}

ostream & operator<<(ostream & os, const Transaction & t) {
	double currentPrice = t.getCurrentPrice(); 
	os << " " << left << setw(5) << t.id << left << setw(10) << t.stock->getName() << left << setw(10) << t.quantity;
	os << left << setw(15) << t.price << left << setw(15) << currentPrice;
	if ((t.price - currentPrice) >= 0) SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), 12);
	else SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), 10);
	os << left << setw(25) << setprecision(6) << abs(currentPrice - t.price) << setprecision(4) << t.getRelativeDiff(currentPrice, t.price) << "%";
	SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), 7);
	return os;
}

Transaction & Transaction::operator=(const Transaction & t) {
	if (this != &t) {
		remove();
		copy(t);
	}
	return *this;
}

Transaction & Transaction::operator=(Transaction && t) {
	if (this != &t) {
		remove();
		move(t);
	}
	return *this;
}

double Transaction::getCurrentPrice() const {
	fstream file;
	string stockName = stock->getName() + ".json";
	file.open(stockName);
	if (!file.is_open()) {
		//time_t seconds = time(NULL);
		//string timestamp = to_string(seconds);
		string timestamp = "1616072670";
		Crawler crawler(stock->getName(), timestamp, timestamp);
		crawler.makeJSON();
	}
	JSONParser jsp(stockName);
	return jsp.getCurrentPrice();
}